File: backend/src/main/java/com/tjost/workouttracker/service/ExerciseService.java
package com.tjost.workouttracker.service;

import org.springframework.transaction.annotation.Transactional;
import com.tjost.workouttracker.exception.ExerciseAlreadyExistsException;
import com.tjost.workouttracker.exception.ExerciseNotFoundException;
import com.tjost.workouttracker.model.Exercise;
import com.tjost.workouttracker.repository.ExerciseRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ExerciseService {

    private static final Logger logger = LoggerFactory.getLogger(ExerciseService.class);
    private final ExerciseRepository exerciseRepo;

    public List<Exercise> getAllExercises() {
        return exerciseRepo.findAll();
    }

    public Exercise getExerciseById(Long exerciseId) {
        return exerciseRepo.findById(exerciseId)
                .orElseThrow(() -> new ExerciseNotFoundException(exerciseId));
    }

    public Exercise createExercise(String name) {
        if (exerciseRepo.existsByNameIgnoreCase(name)) {
            throw new ExerciseAlreadyExistsException(name);
        }

        return exerciseRepo.save(new Exercise(name));
    }

    @Transactional
    public Exercise updateExercise(Long exerciseId, String newName) {
        Exercise exercise = exerciseRepo.findById(exerciseId)
                .orElseThrow(() -> new ExerciseNotFoundException(exerciseId));
        exercise.setName(newName);
        return exercise;
    }

    @Transactional
    public void deleteExercise(Long exerciseId) {
        Exercise exercise = exerciseRepo.findById(exerciseId)
                .orElseThrow(() ->
                        new ExerciseNotFoundException(exerciseId));

        exerciseRepo.delete(exercise);
    }
}

File: backend/src/main/kotlin/workouttracker/backend/service/ExerciseService.kt
package workouttracker.backend.service

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import workouttracker.backend.model.Exercise
import workouttracker.backend.repository.ExerciseRepo
import java.util.*
import kotlin.NoSuchElementException

class ExerciseService(
    private val exerciseRepo: ExerciseRepo) {

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(WorkoutService::class.java)
    }

    fun getAllExercises(): List<Exercise> {
        return exerciseRepo.findAll();
    }

    fun getExerciseById(id: String): Optional<Exercise> {
        return exerciseRepo.findById(id)
    }

    fun getExerciseByName(name: String): Optional<Exercise> {
        if (!exerciseRepo.existsByName(name)) {
            throw NoSuchElementException("No exercise with name $name"); // implement custom exception management later
        }
        return exerciseRepo.findByName(name)
    }

    fun createExercise(exercise: Exercise): Exercise {
        exerciseRepo.save(exercise)
        return exercise
    }

    fun updateExercise(id: String, updated: Exercise): Exercise {
        val existing = exerciseRepo.findById(id)
            .orElseThrow{NoSuchElementException("Exercise with id: $id not found")};
        existing.name = updated.name;
        existing.muscleGroup = updated.muscleGroup;
        existing.weight = updated.weight;
        existing.sets = updated.sets;
        existing.comment = updated.comment;
        return exerciseRepo.save(existing);
    }

    fun deleteExercise(id: String) {
        if (!exerciseRepo.existsById(id)) {
            throw NoSuchElementException("No exercise with id $id"); // implement custom exception management later
        }
        exerciseRepo.deleteById(id);
    }
}
